base/broadcast.jl

```diff
@@ -210,52 +232,52 @@ end
 end
 
 function broadcast_t(f, ::Type{Any}, As...)
-    shape = broadcast_shape(As...)
+    shape = broadcast_indices(As...)
     iter = CartesianRange(shape)
     if isempty(iter)
-        return similar(Array{Union{}}, shape)
+        return similar(Array{Union{}}(0), shape)
     end
     nargs = length(As)
     keeps, Idefaults = map_newindexer(shape, As)
     st = start(iter)
     I, st = next(iter, st)
-    val = f([ As[i][newindex(I, keeps[i], Idefaults[i])] for i=1:nargs ]...)
+    val = f(( _broadcast_getindex(As[i], newindex(I, keeps[i], Idefaults[i])) for i=1:nargs )...)
     B = similar(Array{typeof(val)}, shape)
     B[I] = val
     return _broadcast!(f, B, keeps, Idefaults, As, Val{nargs}, iter, st, 1)
 end
 
-@inline broadcast_t(f, T, As...) = broadcast!(f, similar(Array{T}, broadcast_shape(As...)), As...)
+@inline broadcast_t(f, T, As...) = broadcast!(f, similar(Array{T}, broadcast_indices(As...)), As...)
 
-@inline broadcast(f, As...) = broadcast_t(f, promote_eltype_op(f, As...), As...)
-
-# alternate, more compact implementation; unfortunately slower.
-# also the `collect` machinery doesn't yet support arbitrary index bases.
-#=
-@generated function _broadcast{nargs}(f, keeps, As, ::Type{Val{nargs}}, iter)
+@generated function broadcast_tup{K,ID,AT,nargs}(f, keeps::K, Idefaults::ID, As::AT, ::Type{Val{nargs}}, iter)
     quote
-        collect((@ncall $nargs f i->As[i][newindex(I, keeps[i])]) for I in iter)
+        (((@ncall $nargs f i->_broadcast_getindex(As[i], newindex(I, keeps[i], Idefaults[i]))) for I in iter)...)
```

Comment:
I was wondering if `ntuple` might be more efficient here, but apparently not: #17703